torch/lib/THC/THCTensorIndex.cu

```diff
@@ -332,5 +332,63 @@ __global__ void indexSelectLargeIndex(TensorInfo<T, IndexType> dst,
   }
 }
 
+template <typename IndexType, unsigned int Dims>
+struct LinearIndexCalcData {
+  // sizes for Tensor dims (either from the Tensor, or the size of the adv indexer at that dim)
+  IndexType sizes[Dims];
+  IndexType strides[Dims];       // strides for Tensor
+  bool adv[Dims];                // which Tensors are advanced indexers
+  long *advIndexTensors[Dims];   // Adv Indexing Tensors
+};
+
+template <typename IndexType, unsigned int Dims>
+__device__ __forceinline__ long calculateOffset(
+  IndexType index,                  // index to calculate offset for
+  LinearIndexCalcData<IndexType, Dims> data
+)
+{
+  IndexType offset = 0;
+
+#pragma unroll
+  for (int dim = Dims - 1; dim >= 0; --dim) {
+    IndexType sizeAtDim, strideAtDim, indexAtDim, nextIndex;
+
+    strideAtDim = data.strides[dim];
+    sizeAtDim = data.sizes[dim];
+
+    if (data.adv[dim]) {
+      indexAtDim = data.advIndexTensors[dim][index % sizeAtDim];
+      if (dim > 0 && data.adv[dim - 1]) {
+        nextIndex = index;
+      } else {
+        nextIndex = index / sizeAtDim;
+      }
+    } else {
+      indexAtDim = index % sizeAtDim;
+      nextIndex = index / sizeAtDim;
```

Comment:
Isn't modulo quite expensive on GPUs? I think this would be equivalent:
```cpp
nextIndex = index / sizeAtDim;
indexAtDim = index - nextIndex; // == index % sizeAtDim